packages/website/blog/2023-02-06-consistent-type-exports-and-imports-why-and-how.md

```diff
@@ -0,0 +1,135 @@
+---
+authors:
+  - image_url: https://www.joshuakgoldberg.com/img/josh.jpg
+    name: Josh Goldberg
+    title: typescript-eslint Maintainer
+    url: https://github.com/JoshuaKGoldberg
+description: Why enforcing TypeScript imports use the `type` modifier when possible benefits some project setups.
+slug: consistent-type-imports-and-exports-why-and-how
+tags: [typescript, imports, exports, types, transpiling]
+title: 'Consistent Type Imports and Exports: Why and How'
+---
+
+TypeScript 3.8 [added type-only imports and exports](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export) to the language:
+
+```ts
+import type { SomeThing } from './some-module.js';
+export type { SomeThing };
+```
+
+The key difference with `export type` and `import type` is that they _do not represent runtime code_.
+Attempting to use a _value_ imported as only a _type_ in runtime code will cause a TypeScript error:
+
+```ts twoslash
+import type { SomeThing } from './some-module.js';
+
+new SomeThing();
+//  ~~~~~~~~~
+// 'SomeThing' cannot be used as a value
+// because it was imported using 'import type'.
+```
+
+This is because type-only imports and exports are not emitted as runtime code when code is transpiled to JavaScript.
+This brings up two questions:
+
+- Why would you want to use these type-only imports and exports?
+- How can you enforce a project use them whenever necessary?
+
+Let's Dig In!
+
+<!--truncate-->
+
+## Functional Benefits
+
+### Module Side Effects
+
+Some modules in code may cause _side effects_: code that is run when the module is imported and causes changes outside the module.
+Common examples of side effects include calling to global APIs like `fetch` or creating DOM stylesheets.
+
+When projects include modules that cause side effects, the order of module imports matters.
+For example, some projects import the types of side-effect-causing modules in code that needs to run before those side effects.
+
+### Isolated Module Transpilation
+
+Import statements that only import types are generally removed when the built-in TypeScript compiler transpiles TypeScript syntax to JavaScript syntax.
+The built-in TypeScript compiler is able to do so because it includes a type checker that knows which imports are of types and/or values.
+
+But, some projects use transpilers such as Babel, SWC, or Vite that don't have access to type information.
+These transpilers are sometimes referred to as _isolated module transpilers_ because they effectively transpile each module in isolation from other modules.
+Isolated module transpilers can't know whether an import is of a type, a value, or both.
```

Comment:
The paragraphs above talk about import statements, but the code below is about exports, which is confusing. Generally transpilers _do_ know whether to remove import statements, right?

Reply:
Good point on it being about imports vs exports. They don't know whether to remove imports though, because the seemingly type-only imports might have side effects.

Reply:
> Generally transpilers do know whether to remove import statements, right?

Yes and no. They can make a good guess of it, but it's really hard to know for sure if it's actually safe to remove an import (see halting problem).

Transpilers ofc follow the spec - any top-level type-only import declarations are elided and any inline type-only import specifiers are elided.

Transpilers then make a few assumptions:
- first if a specifier is only used in a type location, then it can be elided.
- if an import without a top-level type-only qualifier has had all of its specifiers elided (either due to the specced behaviour or the assumed behaviour) then the entire import declaration may also be elided.

These assumptions work fine in the vast majority of cases, except in the following:
- decorator metadata can implicitly reference the imported name
- the imported module has side-effects

The entire point of the `verbatimModuleSyntax` flag is that transpilers will make no assumptions at all and will solely elide based on the specced behaviour. Which does mean that:
- zero-specifier import statements will not be elided, just in case there are side-effects
- value specifiers only used in a type location (like classes or enums) that aren't marked with a type-qualifier will not be elided.

Most people won't ever want to use `verbatimModuleSyntax` because the assumptions hold true on their codebase, but some people will need the behaviour (some old angular usecases for example).